pkg/relayer/proxy/websockets/bridge.go

```diff
@@ -0,0 +1,379 @@
+package websockets
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/pokt-network/poktroll/pkg/client"
+	"github.com/pokt-network/poktroll/pkg/observable/channel"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	"github.com/pokt-network/poktroll/pkg/relayer/config"
+	"github.com/pokt-network/poktroll/x/service/types"
+)
+
+// bridge represents a websocket bridge between the gateway and the service backend.
+// It is responsible for forwarding relay requests from the gateway to the service
+// backend and relay responses from the service backend to the gateway.
+type bridge struct {
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+	logger    polylog.Logger
+
+	// serviceBackendConn is the websocket connection to the service backend.
+	serviceBackendConn *connection
+
+	// gatewayConn is the websocket connection to the gateway.
+	gatewayConn *connection
+
+	// msgChan is the channel that the bridge uses to communicate with the connections.
+	msgChan <-chan message
+
+	// relayAuthenticator is the relay authenticator that the bridge uses to verify
+	// relay requests and sign relay responses.
+	relayAuthenticator relayer.RelayAuthenticator
+
+	// relayMeter is the relay meter that the bridge uses to estimate the relay reward
+	// of the bridged relay requests.
+	relayMeter relayer.RelayMeter
+
+	// blockClient is the client used to get the latest block height.
+	blockClient client.BlockClient
+
+	// latestRelayRequest is the latest relay request received from the gateway.
+	// It is used to emit relays to the miner such that there is always a request/response
+	// pair available when submitting proofs.
+	// This is particularly important for asynchronous communication where it is
+	// mostly or exclusively the service backend that sends messages to the gateway
+	// such as eth_subscribe.
+	latestRelayRequest *types.RelayRequest
+
+	// latestRelayResponse is the latest relay response received from the service backend.
+	// It is used to emit relays to the miner such that there is always a request/response
+	// pair available when submitting proofs.
+	// This is particularly important for asynchronous communication where it is
+	// mostly or exclusively the gateway that sends messages to the service backend
+	// such as file upload protocols.
+	latestRelayResponse *types.RelayResponse
```

Comment:
Why don't we use an observable with length 1?

I believe that's what the `Last` function was designed for.

<img width="1285" alt="Screenshot 2025-02-20 at 10 33 12 PM" src="https://github.com/user-attachments/assets/2c717d93-c88c-42c3-81ee-f327197bb24e" />

Reply:
Good question.

I considered that option but it seems to be an unnecessary addition since we don't need to react to `latestRelayResponse` updates.